lib/Sema/CSDiag.cpp

```diff
@@ -1064,16 +1128,24 @@ CalleeCandidateInfo::evaluateCloseness(Type candArgListType,
       // more parameters.
       // We can still do something more sophisticated with this.
       // FIXME: Use TC.isConvertibleTo?
-      if (rArgType->isEqual(paramType))
+
+      SmallVector<ArchetypeType *, 4> archetypes;
+      SmallVector<Type, 4> substitutions;
+      bool matched = findGenericSubstitutions(paramType, rArgType,
+                                              archetypes, substitutions);
+      
+      if (matched && archetypes.size() == 0)
         continue;
-      if (auto genericParam = paramType->getAs<GenericTypeParamType>())
-        paramType = genericParam->getDecl()->getArchetype();
-      if (paramType->is<ArchetypeType>() && !rArgType->hasTypeVariable()) {
+      if (matched && archetypes.size() == 1 && !rArgType->hasTypeVariable()) {
```

Comment:
hasTypeVariable() means the type contains a GenericTypeParamType or a DependentMemberType. I don't believe that can happen here since these are contextual types (written in terms of archetypes) and not interface types (written in terms of type parameters).

Maybe try changing that to an assert. If you're seeing both interface and contextual types here, there might be something else to look at.